hibernate-core/src/main/java/org/hibernate/internal/CacheImpl.java

```diff
@@ -306,7 +309,19 @@ public QueryCache getQueryCache(String regionName) throws HibernateException {
 
 	@Override
 	public void addCacheRegion(String name, Region region) {
-		allCacheRegions.put( name, region );
+		if ( EntityRegion.class.isInstance( region ) ) {
+			entityRegionMap.put( name, (EntityRegion) region );
+		}
+		else if ( CollectionRegion.class.isInstance( region ) ) {
```

Comment:
The instance might implement multiple region types; I'd let it be added to all collections if it matches (dropping the `else`s).

Reply:
I don't think that happens in 5.1 currently, but I see that it could happen. I'll make the suggested change.